log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationSource.java

```diff
@@ -150,23 +141,23 @@ public ConfigurationSource(final InputStream stream) throws IOException {
      * @param data data from the source
      * @param lastModified when the source was last modified.
      */
-    public ConfigurationSource(final Source source, final byte[] data, final long lastModified) {
-        Objects.requireNonNull(source, "source is null");
-        this.data = Objects.requireNonNull(data, "data is null");
-        this.stream = new ByteArrayInputStream(data);
-        this.lastModified = lastModified;
-        this.source = source;
+    public ConfigurationSource(Source source, byte[] data, long lastModified) {
+        this(data, requireNonNull(source, "source is null"), lastModified);
     }
 
-    private ConfigurationSource(final byte[] data, final URL url, final long lastModified) {
-        this.data = Objects.requireNonNull(data, "data is null");
-        this.stream = new ByteArrayInputStream(data);
-        this.lastModified = lastModified;
-        if (url == null) {
-            this.data = data;
-        } else {
-            this.source = new Source(url);
-        }
+    private ConfigurationSource(byte[] data, /*@Nullable*/ Source source, long lastModified) {
+        this(requireNonNull(data, "data is null"), source, new ByteArrayInputStream(data), lastModified);
+    }
+
+    /**
+     * @throws NullPointerException if {@code stream} is {@code null}.
+     */
+    private ConfigurationSource(
+            byte /*@Nullable*/[] data, /*@Nullable*/ Source source, InputStream stream, long lastModified) {
+        this.stream = requireNonNull(stream, "stream is null");
+        this.data = data;
+        this.source = source;
```

Comment:
You state the following:

> `ConfigurationSource` is basically a triple (`Source`, `byte[]` and a `long` timestamp). The `source` and `data` fields can be `null`, but not at the same time.

If so,

1. Would you mind checking here – i.e., at the canonical ctor – both `data` and `source` cannot be null?
2. Can't we do without a `stream` field? Can't we eagerly read the `stream` the first time it is (directly or indirectly via a `Source`) provided? (The only problem is `#getInputStream()`. There we can return `new ByteArrayInputStream(data)`.)

Reply:
I would have to check the code but as I recall the difference in what happens during initial configuration vs reconfiguration is why there are differences in behavior. Before trying to "simplify" you would really have to create a matrix of all the places the ConfigurationSource gets used and determine what data is available and whether the proposed simplifications are safe.

Reply:
1. Fixed in https://github.com/apache/logging-log4j2/pull/2941/commits/3598b6a863dfd6028e72c137936e490ef695a791
2. We could, but this would require a big modification of the behavior: I/O exceptions would appear in the constructors.

Reply:
> We could, but this would require a big modification of the behavior: I/O exceptions would appear in the constructors.

@ppkarwasz, we can wrap them into `UncheckedIOException`s?

Reply:
I agree with @rgoers on this one: I merged the PR and have rolled back the changes to the `ConfigurationSource` file in 755497d3715efe80c543525a0fec3e7171c88dab.